### PR TITLE
Fixed #32820 -- Added aria-invalid="true" to fields with errors.

### PR DESCRIPTION
**Pull request by smithdc1**
Ticket: [#32820](https://code.djangoproject.com/ticket/32820)
Previous PR: https://github.com/django/django/pull/16933

### CODE REVIEW

File: tests/forms_tests/tests/test_formsets.py [hunk 1]
```diff
@@ -1748,9 +1750,9 @@ def test_management_form_invalid_data(self):
             "</ul>"
             "<div>"
             '<input type="hidden" name="form-TOTAL_FORMS" value="two" '
-            'id="id_form-TOTAL_FORMS">'
+            'aria-invalid="true" id="id_form-TOTAL_FORMS">'
```
Comment (smithdc1) [inline tests/forms_tests/tests/test_formsets.py:1753]: @django/accessibility do you think we should still be adding this to "hidden" fields. In this case it's the formset management forms? 

I guess as long as "hidden" fields are not read by screen readers it's ok 🤔

Comment (knyghty) [reply tests/forms_tests/tests/test_formsets.py:1753]: According to https://www.w3.org/TR/html-aria/#input-hidden we shouldn't add any aria attributes to hidden inputs.

File: tests/gis_tests/geoadmin/tests.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ def test_widget_invalid_string(self):
             ' rows="10" name="point" hidden></textarea>',
             output,
         )
-        self.assertEqual(len(cm.records), 1)
+        self.assertEqual(len(cm.records), 2)
```
Comment (smithdc1) [inline tests/gis_tests/geoadmin/tests.py:33]: I'm not sure about this change. It seems wrong to me. Investigating...

Comment (smithdc1) [reply tests/gis_tests/geoadmin/tests.py:33]: See. https://github.com/django/django/pull/17119

So I don't think this is too different to current behaviour. The value is deserialized once for errors and once to render the widget. Errors are cached so would still be two log entries when rendering a form with this patch.

File: django/forms/boundfield.py [hunk 3]
```diff
@@ -105,6 +105,8 @@ def as_widget(self, widget=None, attrs=None, only_initial=False):
             )
         else:
             value = self.value()
+        if not widget.is_hidden and self.errors:
+            attrs["aria-invalid"] = "true"
```
Comment (felixxm) [inline django/forms/boundfield.py:109]: We should probably move this to the `build_widget_attrs()`, check out 3f73df44f2726319f6af44c5613e5e531e9370b6.

File: docs/ref/forms/api.txt [hunk 4]
```diff
@@ -1166,6 +1166,13 @@ Attributes of ``BoundField``
         >>> str(f["subject"].errors)
         ''
 
+    When rendering a field with errors, ``aria-invalid="true"`` will be set on 
+    the field's widget to indicate there is an error to screen reader users.
+
+    .. versionchanged:: 5.0
+
+        The ``aria-invalid="true"`` is set when a field has errors.
```
Comment (felixxm) [inline docs/ref/forms/api.txt:1174]:
```suggestion
        The ``aria-invalid="true"`` was added when a field has errors.
```

File: docs/ref/forms/api.txt [hunk 5]
```diff
@@ -1166,6 +1166,13 @@ Attributes of ``BoundField``
         >>> str(f["subject"].errors)
         ''
 
+    When rendering a field with errors, ``aria-invalid="true"`` will be set on
```
Comment (felixxm) [inline docs/ref/forms/api.txt:1169]:
```suggestion
    When rendering a field with errors, ``aria-invalid="true"`` will be set on
```

File: docs/releases/5.0.txt [hunk 6]
```diff
@@ -310,6 +310,9 @@ Forms
   fields with their help text, the form field now includes the
   ``aria-describedby`` HTML attribute.
 
+* To improve accessibility ``aria-invalid="true"`` is added to the widget of
+  invalid form fields.
```
Comment (felixxm) [inline docs/releases/5.0.txt:314]: Maybe:
```suggestion
* In order to improve accessibility, the invalid form field now includes
  ``aria-invalid="true"`` HTML attribute.
```

File: tests/gis_tests/geoadmin/tests.py (hunk 2 above)
Comment (felixxm) [reply tests/gis_tests/geoadmin/tests.py:33]: Fine for me.

**Review by felixxm — COMMENTED** [review]
@smithdc1 Thanks :+1: I left small suggestions.

**Comment by felixxm** [comment]
> @smithdc1 Thanks +1 I left small suggestions.

@smithdc1 I'd be happy to push edits if you're busy.

File: django/forms/boundfield.py (hunk 3 above)
Comment (smithdc1) [reply django/forms/boundfield.py:109]: Yes. That makes sense. 👍

**Comment by smithdc1** [comment]
> @smithdc1 I'd be happy to push edits if you're busy.

If you could, that would be great. 🥳

**Comment by felixxm** [comment]
@smithdc1 I pushed edits :+1: I will check it again tomorrow.

**Review by felixxm — APPROVED** [review]
@smithdc1 Thanks again :+1: I added `aria-invalid="true"` to HTML outputs in docs.

**Pull request merged**